File: assignment 6/python/programs/5.py
def evaluate_postfix(expression):
    # Initialize an empty stack
    stack = []

    # Iterate through each character in the expression
    for character in expression:
        if character.isdigit():
            # If the character is a digit, push it onto the stack
            stack.append(int(character))
        else:
            # If the character is an operator, pop two operands from the stack and apply the operator
            operand2 = stack.pop()
            operand1 = stack.pop()
            if character == '+':
                result = operand1 + operand2
            elif character == '-':
                result = operand1 - operand2
            elif character == '*':
                result = operand1 * operand2
            elif character == '/':
                result = operand1 / operand2
            elif character == '^':
                result = operand1 ** operand2
            else:
                raise ValueError('Invalid operator')
            # Push the result onto the stack
            stack.append(result)

    # The final result is the only element left in the stack
    return stack[0]

# Example usage
expression = input('Enter a postfix expression: ')
print('Postfix expression:', expression)
print('Result:', evaluate_postfix(expression))

File: assignment 6/c/programs/5.c
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>

#define MAX_SIZE 100

int stack[MAX_SIZE];
int top = -1;

void push(int element)
{
    if (top == MAX_SIZE - 1)
    {
        printf("Stack overflow\n");
        return;
    }
    stack[++top] = element;
}

int pop()
{
    if (top == -1)
    {
        printf("Stack underflow\n");
        return -1;
    }
    return stack[top--];
}

int evaluatePostfix(char *postfixExpression)
{
    int length = strlen(postfixExpression);
    for (int i = 0; i < length; i++)
    {
        char currentCharacter = postfixExpression[i];
        if (isdigit(currentCharacter))
        {
            push(currentCharacter - '0');
        }
        else
        {
            int operand2 = pop();
            int operand1 = pop();
            switch (currentCharacter)
            {
            case '+':
                push(operand1 + operand2);
                break;
            case '-':
                push(operand1 - operand2);
                break;
            case '*':
                push(operand1 * operand2);
                break;
            case '/':
                push(operand1 / operand2);
                break;
            case '%':
                push(operand1 % operand2);
                break;
            default:
                printf("Invalid operator\n");
                return -1;
            }
        }
    }
    return pop();
}

int main()
{
    char postfixExpression[MAX_SIZE];
    printf("Enter a postfix expression: ");
    scanf("%s", postfixExpression);
    int result = evaluatePostfix(postfixExpression);
    if (result != -1)
    {
        printf("The result is: %d\n", result);
    }
    return 0;
}
